### DISCUSSION TITLE
Moonbrooke STGDAT is too short!

### DISCUSSION THREAD

**Discussion by default-kramer** [default-kramer/HermitsHeresy#7] [General]
Moonbrooke has 700 chunks, but the last one is a few bytes short of the normal 0x30000. (Either 544 or 272 bytes, I would need to double-check to be sure.) EDIT: It is 272 bytes, and this bug apparently [applies to Buildertopia islands too!](https://github.com/default-kramer/HermitsHeresy/discussions/21)

On the last chunk, at Y=95, the game stops you from building beyond Z=14,X=26. According to my calculations, there is room in the save file to place blocks up to (but not beyond) Z=27,X=23.

I can't imagine why this special case exists? Who cares about saving 544 bytes in a STGDAT file?

Anyway, I think the best way for Hermit's Heresy to handle this is to just load the last chunk as if it were long enough, allow all modifications as normal, and then truncate the extra bytes off when writing back to the STGDAT file. Seems much better than adding bounds checks all over the place -- once you know the chunk exists, you should be able to assume that all 0x30000 bytes exist.

**Comment by Sapphire645** [comment]
There's a known glitch in that section of the map
https://imgur.com/gallery/dqb2-glitch-island-screenshots-J7YizER
It wasn't known why it happened though, guess we have an answer now; Moonbrooke's file ends early and if something else loads into memory right after it the game just loads it as world data like normal. So there's no special case exception, the game just prays that nothing allocates there. Maybe it's even a bug? Have you tried to load a Moonbrooke file with the full chunk size to see if the game works?

**Reply by default-kramer** [reply]
Ha, very cool to see this! No I have not (yet) tested what happens if I just write the STGDAT file back with the full chunk. But it's on my list.

**Reply by default-kramer** [reply]
But I did just remember, I think the IoA (and most/all the story islands) STGDAT files all uncompress to exactly the same length. Because the IoA has only 369 chunks it just fills zeroes until the end. So possibly they meant to allocate enough space for exactly 700 chunks for all the story islands but just made a math error. As a result, only Moonbrooke is affected because it is the only island that uses all 700 chunks.
